### PR TITLE
feat: message: Add retries to mpool push message from lotus miner

### PR DESCRIPTION
**Pull request by shrenujbansal**
## Related Issues

https://github.com/filecoin-project/lotus/issues/9171

## Proposed Changes

Add retries to mpool push message API in case of unavailability of the lotus chain node

## Checklist

Before you mark the PR ready for review, please make sure that:
- [x] All commits have a clear commit message.
- [x] The PR title is in the form of of `<PR type>: <area>: <change being made>`
    - example: ` fix: mempool: Introduce a cache for valid signatures`
    - `PR type`: _fix_, _feat_, _INTERFACE BREAKING CHANGE_, _CONSENSUS BREAKING_, _build_, _chore_, _ci_, _docs_,_perf_, _refactor_, _revert_, _style_, _test_
    - `area`: _api_, _chain_, _state_, _vm_, _data transfer_, _market_, _mempool_, _message_, _block production_, _multisig_, _networking_, _paychan_, _proving_, _sealing_, _wallet_, _deps_
- [x] This PR has tests for new functionality or change in behaviour
- [x] If new user-facing features are introduced, clear usage guidelines and / or documentation updates should be included in https://lotus.filecoin.io or [Discussion Tutorials.](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [x] CI is green

### CODE REVIEW

File: lib/retry/retry.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	for _, etype := range errorTypes {
+		tmp := reflect.New(reflect.PointerTo(reflect.ValueOf(etype).Elem().Type())).Interface()
+		if errors.As(err, tmp) {
+			return true
+		}
+	}
+	return false
+}
+
+func Retry[T any](attempts int, sleep int, errorTypes []error, f func() (T, error)) (result T, err error) {
```
Comment (magik6k) [inline lib/retry/retry.go:23]: I'd make the `sleep` param be time.Duration.

Comment (magik6k) [reply lib/retry/retry.go:23]: I'd also call the sleep param something that relays that it's increasing - e.g. `initialBackoff`

File: go.mod [hunk 2]
```diff
@@ -38,7 +38,7 @@ require (
 	github.com/filecoin-project/go-fil-commcid v0.1.0
 	github.com/filecoin-project/go-fil-commp-hashhash v0.1.0
 	github.com/filecoin-project/go-fil-markets v1.24.0
-	github.com/filecoin-project/go-jsonrpc v0.1.5
+	github.com/filecoin-project/go-jsonrpc v0.1.7-0.20220818202438-3dabc3d89d31
```
[the remark this answers is not in the document]
Comment (shrenujbansal) [reply go.mod]: done

File: lib/retry/retry.go (hunk 1 above)
Comment (shrenujbansal) [reply lib/retry/retry.go]: done

File: lib/retry/retry.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	"errors"
+	"reflect"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+)
+
+var log = logging.Logger("retry")
+
+func errorIsIn(err error, errorTypes []error) bool {
```
[the remark this answers is not in the document]
Comment (shrenujbansal) [reply lib/retry/retry.go]: done

File: itests/retry_test.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+	"github.com/stretchr/testify/require"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-jsonrpc"
+
+	"github.com/filecoin-project/lotus/lib/retry"
+)
+
+func TestRetryErrorIsInTrue(t *testing.T) {
```
Comment (magik6k) [inline itests/retry_test.go:14]: `itests/` is really just for integration tests which require running test nodes. This test can be in `lib/retry/retry_test.go`; CI will still run that with other unit tests

File: itests/retry_test.go [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+	errorsToRetry := []error{&jsonrpc.RPCConnectionError{}}
+	require.True(t, retry.ErrorIsIn(&jsonrpc.RPCConnectionError{}, errorsToRetry))
```
Comment (magik6k) [inline itests/retry_test.go:16]: Would be nice to also test a wrapped error - something like `require.True(t, retry.ErrorIsIn(xerrors.Errorf("wrapped: %w", &jsonrpc.RPCConnectionError{}), errorsToRetry))`

**Review by magik6k — COMMENTED** [review]
Logic looks good, 2 small comments on the test

**Comment by magik6k** [comment]
There also are some other Go 1.17 references that possibly need updating; this is what seems important on master:

```
testplans/lotus-soup/go.mod
3:go 1.17

testplans/docker-images/Dockerfile.oni-runtime-debug
1:ARG GO_VERSION=1.17.9

testplans/docker-images/Dockerfile.oni-runtime
1:ARG GO_VERSION=1.17.9

testplans/docker-images/Dockerfile.oni-buildbase
1:ARG GO_VERSION=1.17.9

testplans/composer/Dockerfile
1:FROM golang:1.17.9-buster as tg-build

README.md
13:  <a href=""><img src="https://img.shields.io/badge/golang-%3E%3D1.17-blue.svg" /></a>
74:To build Lotus, you need a working installation of [Go 1.17.9 or higher](https://golang.org/dl/):
77:wget -c https://golang.org/dl/go1.17.9.linux-amd64.tar.gz -O - | sudo tar -xz -C /usr/local

Makefile
13:$(error Update Golang to version to at least 1.17.9)

Dockerfile.lotus
1:FROM golang:1.17.9-buster AS builder-deps
```

File: itests/retry_test.go (hunk 5 above)
Comment (shrenujbansal) [reply itests/retry_test.go:16]: good point

File: Makefile [hunk 6]
```diff
@@ -10,7 +10,7 @@ GOCC?=go
 GOVERSION:=$(shell $(GOCC) version | tr ' ' '\n' | grep go1 | sed 's/^go//' | awk -F. '{printf "%d%03d%03d", $$1, $$2, $$3}')
 ifeq ($(shell expr $(GOVERSION) \< 1016000), 1)
 $(warning Your Golang version is go$(shell expr $(GOVERSION) / 1000000).$(shell expr $(GOVERSION) % 1000000 / 1000).$(shell expr $(GOVERSION) % 1000))
-$(error Update Golang to version to at least 1.17.9)
```
Comment (magik6k) [inline Makefile:13]: The `1016000` number above also needs updating (tho it's outdated even on master), I'll handle that to not block this on review anymore

**Review by magik6k — APPROVED** [review]

**Pull request merged**